tests/mutex_test.c

```diff
@@ -0,0 +1,108 @@
+/*
+ *  Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License").
+ *  You may not use this file except in compliance with the License.
+ *  A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+#include <aws/common/mutex.h>
+#include <aws/common/thread.h>
+#include <aws_test_harness.h>
+
+static int test_mutex_acquire_release(struct aws_allocator *allocator, void *ctx) {
+    struct aws_mutex mutex;
+    aws_mutex_init(&mutex, allocator);
+
+    ASSERT_SUCCESS(aws_mutex_acquire(&mutex), "Mutex acquire should have returned success.");
+    ASSERT_SUCCESS(aws_mutex_release(&mutex),"Mutex release should have returned success.");
+
+    aws_mutex_clean_up(&mutex);
+
+    return 0;
+}
+
+struct thread_mutex_data {
+    struct aws_mutex mutex;
+    volatile int counter;
+    int max_counts;
+    volatile int thread_fn_increments;
+};
+
+static void mutex_thread_fn(void *mutex_data) {
+    struct thread_mutex_data *p_mutex = (struct thread_mutex_data *)mutex_data;
+    int finished = 0;
+    while (!finished) {
+        aws_mutex_acquire(&p_mutex->mutex);
+        if (p_mutex->counter != p_mutex->max_counts) {
+            int counter = p_mutex->counter + 1;
+            p_mutex->counter = counter;
+            p_mutex->thread_fn_increments += 1;
+            finished = p_mutex->counter == p_mutex->max_counts;
+        }
+        else {
+            finished = 1;
+        }
+        aws_mutex_release(&p_mutex->mutex);
+
+    }
+}
+
+static int test_mutex_is_actually_mutex(struct aws_allocator *allocator, void *ctx) {
+
+    struct thread_mutex_data mutex_data = {
+        .counter = 0,
+        .max_counts = 1000000,
+        .thread_fn_increments = 0,
+    };
+
+    aws_mutex_init(&mutex_data.mutex, allocator);    
+
+    struct aws_thread thread;
+    aws_thread_init(&thread, allocator);
+    ASSERT_SUCCESS(aws_thread_create(&thread, mutex_thread_fn, &mutex_data, 0), "thread creation failed with error %d", aws_last_error());
+    int finished = 0;
+    int increments = 0;
+    while (!finished) {
+        aws_mutex_acquire(&mutex_data.mutex);
+        /*add some fairness for thread startup time.*/
```

Comment:
Not sure I understand what this comment is referring to.

Reply:
If thread takes too long to startup, the test doesn't acquire the lock very many times. This is added to make sure thread 1 doesn't do all the work before thread 2 gets a chance.

Reply:
You're just acquiring and releasing here, so if thread 1 starts up before your acquire you still have this problem. Acquire the mutex before launching any threads if you want to do this.

Reply:
thread 1 is already started (it's the main thread). This will keep going until thread 2 acquires the lock once and then they start competing for the lock.